fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneConcurrency.java

```diff
@@ -0,0 +1,162 @@
+     * @param async the future to wait on
+     * @param recordContext the context to use for callback, recording the event and getting the timeout
+     *
+     * @return the result of the future's operation
+     */
+    @Nullable
+    @API(API.Status.INTERNAL)
+    public static <T> T asyncToSync(@Nonnull StoreTimer.Wait event, @Nonnull CompletableFuture<T> async, @Nonnull FDBRecordContext recordContext) {
+        if (recordContext.getPropertyStorage().getPropertyValue(LuceneRecordContextProperties.LUCENE_USE_LEGACY_ASYNC_TO_SYNC)) {
+            return recordContext.asyncToSync(event, async);
+        }
+
+        if (recordContext.hasHookForAsyncToSync() && !MoreAsyncUtil.isCompletedNormally(async)) {
+            recordContext.getHookForAsyncToSync().accept(event);
+        }
+
+        recordContext.getDatabase().checkIfBlockingInFuture(async);
+
+        if (async.isDone()) {
+            try {
+                return async.get();
+            } catch (ExecutionException ex) {
+                throw FDBExceptions.wrapException(ex);
+            } catch (InterruptedException ex) {
+                Thread.currentThread().interrupt();
+                throw FDBExceptions.wrapException(ex);
+            }
+        } else {
+            final Pair<Long, TimeUnit> asyncToSyncTimeout = recordContext.getDatabase().getAsyncToSyncTimeout(event);
+            Timeout timeout = (asyncToSyncTimeout == null) ? null : Timeout.of(asyncToSyncTimeout.getLeft(), asyncToSyncTimeout.getRight());
+            final FDBStoreTimer timer = recordContext.getTimer();
+            final long startTime = System.nanoTime();
+            try {
+                if (timeout != null) {
+                    return async.get(timeout.getDuration(), timeout.getTimeUnit());
+                } else {
+                    return async.get();
+                }
+            } catch (TimeoutException ex) {
+                if (timer != null) {
+                    timer.recordTimeout(event, startTime);
+                    throw new AsyncToSyncTimeoutException(ex.getMessage(), ex,
+                            LogMessageKeys.TIME_LIMIT.toString(), timeout.getDuration(),
+                            LogMessageKeys.TIME_UNIT.toString(), timeout.getTimeUnit());
+                }
+                throw new AsyncToSyncTimeoutException(ex.getMessage(), ex);
+            } catch (ExecutionException ex) {
+                throw FDBExceptions.wrapException(ex);
+            } catch (InterruptedException ex) {
+                Thread.currentThread().interrupt();
+                throw FDBExceptions.wrapException(ex);
+            } finally {
+                if (timer != null) {
+                    timer.recordSinceNanoTime(event, startTime);
+                }
+            }
+        }
+    }
+
+    private LuceneConcurrency() {
+    }
+
+    /**
+     * A representation of a timeout value: A duration and a time unit.
+     */
+    public static class Timeout {
```

Comment:
I am not exactly opposed to the creation of this new class, but I see that the only(?) usage constructs one from separate values retrieved via an API that returns `Pair` and then deconstructs it just a few lines later.

Would it make more sense to push this into `getAsyncToSyncTimeout` in place of the `Pair`? I guess that's an incompatible change.

Would it make sense to also include the `Event`, so that it's more scoped to the abstracts immediately here?

Reply:
The original idea behind this class was the ability to add a version of `asyncToSync` that could take the timeout value without consulting the `FDBDatabase`, but that did not pan out because it is incompatible with existing tests and would not end up being useful. Then I left the class in place as it offered some indication of the constituents of the timeout (as opposed to `Pair.getRight` and `Pair.getLeft`), but if that is causing confusion, I will remove it.

Reply:
Done (Removed `Timeout` class)